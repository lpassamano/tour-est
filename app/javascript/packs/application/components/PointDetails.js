import React, { Component } from "react";
import PropTypes from "prop-types";
import { MediaObject, MediaObjectSection, Thumbnail } from "react-foundation";
import { Link } from "@reach/router";
import { connect } from "react-redux";
import * as pointActions from "../redux/points/actions";
import * as pointSelectors from "../redux/points/selectors";

export class PointDetails extends Component {
  // on refresh there is no stored state!
  // so we can't search state to get current point!!!

  componentDidMount() {
    this.props.getPoint(this.props.tourId, this.props.pointId);
    this.props.listPoints(this.props.tourId);
  }

  render() {
    const { isFetching, point, tourId, nextPoint, previousPoint } = this.props;

    if (isFetching || !point) {
      return <p>loading... please wait!</p>;
    }

    const {
      id,
      title,
      image,
      caption,
      description,
      location,
      directions
    } = point;

    return (
      <div key={id}>
        <div className="media-object-basics-example">
          <h4>{title}</h4>
          {image ? (
            <MediaObject stackForSmall>
              <MediaObjectSection>
                <Thumbnail src={image} alt=" " width="150" />
              </MediaObjectSection>
              <MediaObjectSection isMain>
                {caption && <p>{caption}</p>}
                {location && <p>Location: {location}</p>}
                {directions && <p>How to get there: {directions}</p>}
                {description && <p>{description}</p>}
              </MediaObjectSection>
            </MediaObject>
          ) : (
            <div>
              {caption && <p>{caption}</p>}
              {location && <p>Location: {location}</p>}
              {directions && <p>How to get there: {directions}</p>}
              {description && <p>{description}</p>}
            </div>
          )}
          <hr />
          {nextPoint ? (
            <Link
              to={`/tours/${tourId}/points/${nextPoint.id}`}
              className="button expanded"
            >
              Next
            </Link>
          ) : (
            <Link to={`/tours/${tourId}`} className="button expanded">
              Tour Over! <br />
              Back to Tour Info
            </Link>
          )}
          <div className="button-group expanded">
            {previousPoint && (
              <Link
                to={`/tours/${tourId}/points/${previousPoint.id}`}
                className="button hollow"
              >
                Previous Point
              </Link>
            )}
            {nextPoint && (
              <Link to={`/tours/${tourId}`} className="button hollow">
                Back to Tour Info
              </Link>
            )}
          </div>
        </div>
      </div>
    );
  }
}

const pointProps = PropTypes.shape({
  title: PropTypes.string.isRequired,
  caption: PropTypes.string,
  description: PropTypes.string,
  location: PropTypes.string,
  directions: PropTypes.string,
  image: PropTypes.string
});

PointDetails.propTypes = {
  point: pointProps,
  isFetching: PropTypes.bool.isRequired,
  getPoint: PropTypes.func.isRequired,
  listPoints: PropTypes.func.isRequired,
  nextPoint: pointProps,
  previousPoint: pointProps
};

const mapStateToProps = (state, ownProps) => ({
  point: pointSelectors.getPoint(state, ownProps.pointId),
  isFetching: pointSelectors.isFetching(state),
  nextPoint: pointSelectors.getNextPoint(state, ownProps.pointId),
  previousPoint: pointSelectors.getPreviousPoint(state, ownProps.pointId)
});

const mapDispatchToProps = {
  listPoints: pointActions.listPoints,
  getPoint: pointActions.getPoint
};

const enhance = connect(
  mapStateToProps,
  mapDispatchToProps
);

export default enhance(PointDetails);
