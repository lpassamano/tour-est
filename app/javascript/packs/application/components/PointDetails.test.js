import { PointDetails } from "./PointDetails";
import React from "react";
import { shallow } from "enzyme";

describe("<PointDetails />", () => {
  const setup = () => {
    const point = {
      title: "First point"
    };
    const isFetching = false;
    const listPoints = jest.fn();
    const getPoint = jest.fn();
    const component = shallow(
      <PointDetails
        point={point}
        listPoints={listPoints}
        isFetching={isFetching}
        getPoint={getPoint}
      />
    );
    return component;
  };

  test("render", () => {
    const component = setup();
    expect(component).toMatchSnapshot();
  });
});
