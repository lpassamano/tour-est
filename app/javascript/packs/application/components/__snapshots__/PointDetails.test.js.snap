// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PointDetails /> render 1`] = `
<div>
  <div
    className="media-object-basics-example"
  >
    <h4>
      First point
    </h4>
    <div />
    <hr />
    <[object Object]
      className="button expanded"
      to="/tours/undefined"
    >
      Tour Over! 
      <br />
      Back to Tour Info
    </[object Object]>
    <div
      className="button-group expanded"
    />
  </div>
</div>
`;
